src/wasm-traversal.h

```diff
@@ -157,185 +149,209 @@ struct WasmWalkerBase : public WasmVisitor<SubType, ReturnType> {
     self->visitMemory(&module->memory);
     self->visitModule(module);
   }
-};
 
-template<typename ParentType>
-struct ChildWalker : public WasmWalkerBase<ChildWalker<ParentType>> {
-  ParentType& parent;
+  // Walk implementation. We don't use recursion as ASTs may be highly
+  // nested.
 
-  ChildWalker(ParentType& parent) : parent(parent) {}
+  // Tasks receive the this pointer and a pointer to the pointer to operate on
+  typedef void (*TaskFunc)(SubType*, Expression**);
 
-  void visitBlock(Block *curr) {
-    ExpressionList& list = curr->list;
-    for (size_t z = 0; z < list.size(); z++) {
-      parent.walk(list[z]);
-    }
-  }
-  void visitIf(If *curr) {
-    parent.walk(curr->condition);
-    parent.walk(curr->ifTrue);
-    parent.walk(curr->ifFalse);
-  }
-  void visitLoop(Loop *curr) {
-    parent.walk(curr->body);
-  }
-  void visitBreak(Break *curr) {
-    parent.walk(curr->condition);
-    parent.walk(curr->value);
-  }
-  void visitSwitch(Switch *curr) {
-    parent.walk(curr->condition);
-    if (curr->value) parent.walk(curr->value);
-  }
-  void visitCall(Call *curr) {
-    ExpressionList& list = curr->operands;
-    for (size_t z = 0; z < list.size(); z++) {
-      parent.walk(list[z]);
-    }
-  }
-  void visitCallImport(CallImport *curr) {
-    ExpressionList& list = curr->operands;
-    for (size_t z = 0; z < list.size(); z++) {
-      parent.walk(list[z]);
-    }
+  struct Task {
+    TaskFunc func;
+    Expression** currp;
+    Task(TaskFunc func, Expression** currp) : func(func), currp(currp) {}
+  };
+
+  std::vector<Task> stack;
+
+  void addTask(TaskFunc func, Expression** currp) {
```

Comment:
Should we call this pushTask to make the stacking behavior a little more explicit?